### ISSUE TITLE
Preferred Name Order: Given Family vs Family Given

### ISSUE THREAD

**Issue by bmcutler** [Submitty/Submitty#8446]
**What problem are you trying to solve with Submitty**
When Submitty combines given & family names, we currently assumes "Given Family" is the correct order.  But the preferred order of names varies in different areas of the world -- users may prefer the combined order "Family Given". 

**Describe the way you'd like to solve this problem**
Allow students to specify in the profile which order their names should appear, when we display a combined name.

**Describe any potential alternatives you'd tried to solve the problem**
We are not currently storing the middle name.  This may impact the flexibility in constructing the correct combined name.  We may not be able to just assume that attaching the middle name after the given name is appropriate.  E.g.,  "Given Middle Family" vs "Family Given Middle" vs. "Family Middle Given".
Adding a middle name is probably not necessary for teaching & grading.  And we probably don't have a source of data for this (our direct registrar registration input does not include middle name).

**Comment by Smeyer025** [comment]
Hi! I'm new to open source, and I would like to try to solve this issue. Do I need to be assigned the issue, or can I just get started?

**Comment by LeoJung1** [comment]
Hi, is this issue open for change? If so, we would like to work on this issue as a team along with issue #8436!

**Comment by abhishekyakhmi** [comment]
I would like to work on this issue  , if it's still open.

**Comment by pranavsambyal** [comment]
Hey, I would like to propose a solution , that is to use ask the user to choose a display  name while signup which an be used. This avoids implementing a toggle switch for all  users preferred orders.

**Comment by VaibhavKambar7** [comment]
Has this issue been solved?

**Comment by sanyamk30** [comment]
Hey @bmcutler , is the issue still open? I would like to work on it

**Comment by ziesski** [comment]
I reviewed PR #10259 today, and while golfing I was thinking about this PR and the design decision, and personally think this is NOT the way we should solve the problem.

It would be confusing if we allow users to decide which order individually. Usually it is obvious enough to tell which is Given and which is Family, but it is also common for people to have interchangeable Given and Family name. This would be hard to tell for users (and even worse for international students where English is not their first language). 

I rather would propose it to be set in submitty.json during initial configuration process, and if Submitty is used in different cultural location, admin can decide which order to use. 

If current implementation is applied to forum, it would possible make giant mess, and I think it is over empowering users. We should stick to the convention, and keep it consistent throughout the course. 

@bmcutler @micpap25

**Comment by micpap25** [comment]
@ziesski I disagree with "Usually it is obvious enough to tell which is Given and which is Family". I also think it's common to have a mix of students of different cultures who want to personally customize the order.

I don't believe this is over-empowering users, and I think like pronouns, students should be fully able to choose how they will be addressed in the forum space. This is especially true as options like a live lecture chat are being considered.

With that said I think this feature has a long way to go to be implemented, and I'm happy to discuss what needs to be done to ensure students are able to properly express themselves while still minimizing confusion for instructors / TAs

**Comment by micpap25** [comment]
After speaking to Jaeseok I agree, I think we should rely on preferred first / last names when allowing users to set the way they will be referred to, remove references to given / family name in the context of first / last name, and have a setting on Submitty which determines when importing from the Registrar how the name in Submitty will be determined from registrar data. (I believe registrar uses given / family, so the submitty config can allow users to match given/family to first/last or given/family to last/first.)

**Comment by ziesski** [comment]
Yes, I also understood Michael's point, and do not think it is over empowering users, because there are occasions where user strongly feels the need to edit their name, and how they are displayed in forum, and we both agreed on utilizing existing preferred name setting to handle such case. 
![스크린샷 2024-07-08 오전 11 05 21](https://github.com/Submitty/Submitty/assets/123261952/a65608ce-2dbb-478d-8bbb-b1251b1359cf)

For future reference, during the conversation, I was using google profile page as an example.
Notice how I have Korean account and American account, and the name order is set. Although I could edit the name, I do not have control over which order it is displayed. It is just set as convention. I guess if user is strongly willing to edit the order, they could put their first name in the last name section, and vice versa. 

![스크린샷 2024-07-08 오전 11 00 51](https://github.com/Submitty/Submitty/assets/123261952/c5b08a50-45c7-4d1b-bf99-bdf79d3e7fed)
![스크린샷 2024-07-08 오전 11 01 10](https://github.com/Submitty/Submitty/assets/123261952/18fd8a5c-b889-49bf-91c8-2a1de722c154)

**Comment by bmcutler** [comment]
@micpap25 @ziesski 

Respectfully I disagree with the direction of this conversation.
We want to move away from the using "first"/"last" everywhere in the code and on the UI -- this is non-inclusive terminology.
We want to use "given" and "family".
While a university can set a default based on the majority of their students (this would be an additional issue/PR, an additional server setting in the config folder), individuals should be able to specify an ordering different from the university default to welcome students from any country/culture.   

We do NOT want students who are in the minority swapping their family & given names just so they display in the proper order.  This HACK will cause no end of headaches and confusion when faculty attempt to send alerts or enter final grades -- students will be alphabetized differently and in the worst case the wrong grade will be submitted for students.

I give as an example the way olympic figure skating results are displayed (from way back in 1998):

<img width="500" alt="Screenshot 2024-07-08 at 10 02 05 PM" src="https://github.com/Submitty/Submitty/assets/5871417/2b3ca727-64f0-464d-80ec-73b5aafe50d4">

It is awesome when sports announcers and broadcasters respect the athletes and use their preferred name order.
To help highlight the ordering, sometimes the TV display for sports will use ALL CAPS to highlight the family names.
I can't quickly find an example now to screenshot, but it would look like:

Tara LIPINSKI
Michelle KWAN
CHEN Lu

I believe we always want to alphabetize by family name first, then given name.
So yes, in a class with many students with the family name Chen, some who prefer "family given" and others who prefer "given family" the lab attendance / check point sheet might look like:

Alice CHEN
CHEN Fu
CHEN Mei
Robert CHEN
CHEN Xiang
Robert SMITH
WANG Chun
Robert WANG

**Comment by ziesski** [comment]
@bmcutler  I do understand and agree we want to respect the cultural sensitivity, and yes the hacky way of swapping can cause serious mistakes. 
However, I have some concerns, for example, the screenshot of 1998 winter olympics non-english speaker could so easily get confused, because while Chen is very common family name, it can also be common given name, and [Lu](https://en.wikipedia.org/wiki/Lu_(surname)) is also common family name. So typical non-native speaker might go series of deduction -> Michelle seems like a given name, and Lipinski seems like a family name so Chen could be the given name. 

![스크린샷 2024-07-08 오후 11 47 36](https://github.com/Submitty/Submitty/assets/123261952/6fcf7a4c-591e-479b-b916-268dd8bed6b8)
Here is my example, which is from 2009 world figure skating championship Ladies' final standings.
Notice how Kim Yuna and Mao Asada's name are oriental, but Ando Miki's name is not, and this could cause confusion if we do not make sure things are applied consistently. 

However, I like the idea of making the family name all uppercase letters. Seems like an intuitive way to tell which is given/family. 

Few concerns may be corner cases may include case sensitive Family name such as "McCarthy" which is conventional way to spell names of Gaelic origin. 

Or SQL / JS sorting methods, Some SQL sorts by A-Z-a-z and some does it A-a-Z-z, I think PostgreSQL is latter. It may need some major overhaul depending on current implementation.

**Comment by bmcutler** [comment]
In addition to the all caps, I have also see displays with the last name is in bold font, so it stands out, especially in a list with many people with different preferred ordering.

Alice **CHEN**
**CHEN** Fu
**CHEN** Mei
Robert **CHEN**
**CHEN** Xiang
Robert **SMITH**
**WANG** Chun
Robert **WANG**

Yes, by using all caps we sacrifice the mixed capitalization in names like McCarthy.
BTW: alphabetization of Irish names, e.g., McCarthy and MacCarthy, often considers Mc and Mac to be equivalent.
See https://en.wikipedia.org/wiki/Mac_and_Mc_together

Also, we should consider multi word names with varied capitalization and spaces (e.g., Spanish names) and how that impacts and the capitalization and spaces and alphabetization.
See https://en.wikipedia.org/wiki/Spanish_naming_customs

On multiple occasions I have seen minor variations in the sorting used by the RPI data systems and the Submitty sorting when entering final grades.  Instructors just have to be REALLY careful when walking through a lengthy list of students.

Yes, after looking at a single page/instance of a name some readers may not instantly know which part of the name is the family name and may be confused or may make incorrect first assumptions, that's ok, that happens all the time in real life.  

The long term plan should be:

1) Replace all uses of "first" and "last" with "given" and "family" within our UI and code.

2) Consider adding middle name field to the database.  Do we also add a preferred middle name column?  Make sure we match legal name storage conventions, and make sure we can import/export names to/from university data sources.  Note that even if we get middle name data from a university source, many students will likely prefer not to include that name in the display of their "full" name for class purposes.  

3) ask students how their full name should be displayed - e.g., when the instructor or TA or grader sees their complete name.  One clear example for display of the preferred full name order is when looking at the short list of students who are on a team:
    <img width="521" alt="Screenshot 2024-07-09 at 12 47 29 PM" src="https://github.com/Submitty/Submitty/assets/5871417/ee917616-eedd-42c3-80bf-6686939aab68">

4) make sure instructors/TAs/graders have some control over displays of lengthy lists of students, e.g., for attendance grading, so that it is intuitive what order the students are listed so they can find specific students.  Allowing the instructor/TA/grader to choose a) preferred full name ordering with capitalization/bold to highlight family names or b) break the display into separate columns with uniform ordering (ignoring the students preferred ordering).
    <img width="546" alt="Screenshot 2024-07-09 at 12 47 57 PM" src="https://github.com/Submitty/Submitty/assets/5871417/1f4fb013-00a3-42d8-a31d-2d3dbc8352fc">
    NOTE: we need to replace first/last with given/family in this screenshot!

5) ask students how a shortened version of their name should be displayed to their peers within the discussion forum - e.g., a) just their given name or b) their given name +  first letter of their family name or c) family name initial + given name or d) full name with preferred ordering of given & family or ...  any other options that make sense to add to this list.  The choice depends on how large the class is, how unique or common their given name is, and what level of anonymity/privacy the student wants to have.  NOTE: Some of this work was done in https://github.com/Submitty/Submitty/pull/9226, but I think this UI could be made more intuitive and the list of choices expanded.

6) consider adding some server configuration to specify a default name display order choice that matches the majority of students at the university for convenience.
